portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -116,18 +137,53 @@ private int getMaxNumComments(HttpServletRequest request) {
     return maxComments;
   }
 
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) return null;
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // To support running in Google Cloud Shell with AppEngine's devserver, we must use the relative
+    // path to the image, rather than the path returned by imagesService which contains a host.
+    try {
+      URL url = new URL(imagesService.getServingUrl(options));
+      return url.getPath();
+    } catch (MalformedURLException e) {
+      return imagesService.getServingUrl(options);
+    }
+  }
+
   /**
-   * Converts an ArrayList<String> into a JSON string.
+   * Converts a list of comments into a JSON string.
    */
-  private String convertToJson(ArrayList<String> comments) {
+  private String convertToJson(ArrayList<ArrayList<String>> comments) {
     String json = "{";
     json += "\"history\": [";
     for (int i = 0; i < comments.size(); i++) {
-      String comment = comments.get(i);
-      if (i != 0) {
-        json += ", ";
-      }
-      json += "\"" + comment + "\"";
+      ArrayList<String> commentWithImage = comments.get(i);
+      if (i != 0) json += ", ";
+      json += "[\"" + commentWithImage.get(0) + "\", \"";
+      String imageUrl = commentWithImage.get(1);
+      if (imageUrl == null) json += "";
+      else json += imageUrl;
```

Comment:
Why not invert the condition to avoid adding nothing to json?

"if (imageUrl != null) json += imageUrl;"